### PR TITLE
Add delegate notification for missing competition results after competition is over, fixes #2116

### LINKED ISSUE #2116: Add a notification for competitions that are over, but the delegate has not yet submitted results for
I discovered that we don't actually have that over in our [notifications_helper](https://github.com/thewca/worldcubeassociation.org/blob/master/WcaOnRails/app/helpers/notifications_helper.rb) while investigating #2113 with @SqAtx.

This might make for another good ment2 for @SqAtx when he's done with #2113 =)

### PR DESCRIPTION
**Pull request by LinusFresz**
Fixes https://github.com/thewca/worldcubeassociation.org/issues/2116.

I've thought about another link for the new notification instead of the competition website but as long as https://github.com/thewca/worldcubeassociation.org/issues/2113 isn't done I don't think we can link to anywhere useful.

Is is better two have two separate notifications for results and reports? I personally like the combination of both.

EDIT: Sorry for the mess with the commits, still figuring out Git. :/

### CODE REVIEW

File: WcaOnRails/app/helpers/notifications_helper.rb [hunk 1]
```diff
@@ -68,13 +68,24 @@ def notifications_for_user(user)
     end
 
     user.delegated_competitions.visible.over
-        .includes(:delegate_report).where(delegate_reports: { posted_at: nil })
-        .each do |competition|
+      .each do |competition|
```
Comment (jfly) [inline WcaOnRails/app/helpers/notifications_helper.rb]: I'm pretty sure that removing the `.includes(:delegate_report)` introduces an N+1 query here (see https://secure.phabricator.com/book/phabcontrib/article/n_plus_one/ for an explanation).

**Comment by jfly** [comment]
Also, looks like /spec/helpers/notifications_helper_spec.rb is failing tests. Hopefully straightforward to fix.

File: WcaOnRails/app/helpers/notifications_helper.rb [hunk 2]
```diff
@@ -68,13 +68,24 @@ def notifications_for_user(user)
...
+        if !competition.delegate_report.posted_at && !competition.results_posted_at?
```
Comment (jonatanklosko) [inline WcaOnRails/app/helpers/notifications_helper.rb]: We have some nice boolean methods for this: `!competition.delegate_report.posted?` and `!competition.results_posted?`.

**Comment by jonatanklosko** [comment]
Concerning the merge commit: you don't have to merge `upstream/master` unless there are some conflicts that need to be resolved, in which case we usually do `git rebase` instead =)

**Comment by LinusFresz** [comment]
Well but can it already be used? :-) As far as I see it is not on production and ready for use yet?

**Comment by jfly** [comment]
It is live! I don't think we've added any links to it yet, but you can see it for any competition that is in the past and does not have results uploaded yet, such as https://staging.worldcubeassociation.org/competitions/XuzhouOpen2018/submit-results.

File: WcaOnRails/app/helpers/notifications_helper.rb (hunk 1 above)
Comment (LinusFresz) [reply WcaOnRails/app/helpers/notifications_helper.rb]: If I am using .includes(:delegate_report) then I get some notification:

user: lf
AVOID eager loading detected
  Competition => [:delegate_report]
  Remove from your finder: :includes => [:delegate_report]
[...]

In my noobish imagination I assumed this is worse than going without .includes(:delegate_report) which does not provide any messages. Am I wrong?

File: WcaOnRails/app/helpers/notifications_helper.rb [hunk 3]
```diff
@@ -68,13 +68,24 @@ def notifications_for_user(user)
...
           notifications << {
+            text: "The delegate report and competition results for #{competition.name} have not been submitted.",
+            url: delegate_report_path(competition),
```
Comment (jfly) [inline WcaOnRails/app/helpers/notifications_helper.rb]: I don't know that it makes sense to link to the delegate report here.... could you explain your reasoning? I feel like it should be fine to link to the competition, and we should make sure that competition page has useful warnings and links to take care of the delegate report and submitting results.

Comment (LinusFresz) [reply WcaOnRails/app/helpers/notifications_helper.rb]: After reviewing it seems even more useful to link to the results page instead. But I guess you are right, a delegate should get all necessary information from the competition's main page (therefore a message should show up which indicates the responsible delegates to submit the results).

Comment (jfly) [reply WcaOnRails/app/helpers/notifications_helper.rb]: Does our main page show useful information for a delegate who has work to do? If not, could you file an issue with some ideas for improvements? =)

Comment (LinusFresz) [reply WcaOnRails/app/helpers/notifications_helper.rb]: I meant the competition main page, such as https://www.worldcubeassociation.org/competitions/MunichOpen2018
There is already a notification for missing delegate reports, so why not add another one for missing competition results?! :)

File: WcaOnRails/app/helpers/notifications_helper.rb (hunk 1 above)
Comment (jfly) [reply WcaOnRails/app/helpers/notifications_helper.rb]: Damn, what a pain =( We have this tool (bullet) that tries to autodetect possible N+1 issues, but I'm not sure how accurate it actually is.

When running the server, you should see a list of all SQL queries that get generated when viewing a page. Can you try both with and without the `.includes` and see which queries end up getting generated? Hopefully it will be straightforward from looking at that which is the right way to go. If it turns out that `bullet` is wrong, we can teach it to ignore this query (see https://github.com/jfly/worldcubeassociation.org/blob/866eab8597685c45fb06d6847c0a118ec84b662f/WcaOnRails/config/environments/development.rb#L76-L82)

Comment (LinusFresz) [reply WcaOnRails/app/helpers/notifications_helper.rb]: OK I've looked through the requests and in both ways actually N+1 queries are getting generated.. :/ I assume this is not created by removing `.includes(:delegate_report)` but using if-else cases what calls each competition separate anyway. 

I can avoid this by looking at results and delegate reports in *two separate loops* but I would have liked to do it in one run.

**Comment by LinusFresz** [comment]
I've updated according to your comments @jfly @jonatanklosko 
Of course I removed the N+1 query which resulted in some other problems: I've had trouble fixing the notifications_helper_spec.rb and I don't exactly know why. It's pretty clear that during the test the results for the delegate report test and the competition result tests get merged, but why? Because of the similar specs for the competition without results/report?

File: WcaOnRails/app/views/competitions/_nav.html.erb [hunk 4]
```diff
@@ -187,6 +187,14 @@
         <% end %>
       <% end %>
 
+      <% if @competition.user_should_post_competition_results?(current_user) %>
+        <%= alert :warning, note: true do %>
+          The results are not submitted yet! Click
```
Comment (jfly) [inline WcaOnRails/app/views/competitions/_nav.html.erb]: nit: Notice how the delegate report warning is personal? It uses the word "your", which I think gives the reader a clear sense that this is something that is their responsibility.

I might consider rewording this a bit. Maybe this: "Please submit results for this competition. You can do so " ?

Whatever you think is best!

**Comment by LinusFresz** [comment]
Should be good now! Do I have to write a test for the newly created notification on the competition page? ('Please submit results for this competition!')
Not sure if this is tested as well! :-)

File: WcaOnRails/app/views/competitions/_nav.html.erb [hunk 5]
```diff
@@ -187,6 +187,12 @@
         <% end %>
       <% end %>
 
+      <% if @competition.user_should_post_competition_results?(current_user) %>
+        <%= alert :warning, note: true do %>
+          Please submit results for this competition! You can upload them          <%= link_to "here", submit_results_edit_path(@competition) %>.
```
Comment (jfly) [inline WcaOnRails/app/views/competitions/_nav.html.erb]: nit: There's a lot of unnecessary whitespace in the middle of this line.
